### ISSUE TITLE
Feature suggestion: LAST_INSERT_UUID()

### ISSUE THREAD

**Issue by gitasmus** [dolthub/dolt#7547]
In his blog Tim pointed out the usefulness of UUIDs: https://www.dolthub.com/blog/2023-10-27-uuid-keys/

A minimal example could look as follows. Create table with a UUID as default primary key:
```SQL
CREATE TABLE uuid_test(
    id BINARY(16) DEFAULT (UUID_TO_BIN(UUID())),
    some_text VARCHAR(255),
    PRIMARY KEY (id)
);
```
Next Insert a row without giving an ID:
```SQL
INSERT INTO uuid_test (some_text) VALUES ('abc');
```
If we are interested in the last inserted ID 
```SQL
SELECT LAST_INSERT_ID();
``` 
does not return the UUID as it's return type is hard-wired to `UNSIGNED BIGINT`.

In analogy to `LAST_INSERT_ID()` it would thus be nice if _dolt_ provided a feature that is _also_ missing in _MySQL_: `LAST_INSERT_UUID()`, which would retur the last auto-created UUID.

The lack of such a `LAST_INSERT_UUID()` function could be circumvented in _MySQL_ and _dolt_ with a `TRIGGER` as described in 
https://stackoverflow.com/questions/4687196/mysql-insert-id-or-something-like-that-to-return-last-mysql-uuid

A dedicated `LAST_INSERT_UUID()` function in dolt would IMHO be appropriate to the importance of UUIDs (and convenient), but would break the compatibility with _MySQL_. The alternative of implementing the return type of `LAST_INSERT_ID()` in a more general way to allow for `ID` to be `UNSIGNED BIGINT`, `UUID()` to be `VARCHAR(36)` and `UUID_TO_BIN(UUID())` to be `BINARY(16)` could be even more logical but would cause even more compatibility confusion with _MySQL_.

Perhaps, it would be the better approach to ask for such a feature in _MySQL_ first and then ask for the compatibility of _dolt_ here ...

**Comment by timsehn** [comment]
This is a good feature request. I think it's fine for us to extend MySQL. We'll have someone implement this function today. Likely @fulghum.

**Comment by fulghum** [comment]
Thanks for the great feature request @gitasmus! 🙏  I'm getting started on this one now for you.

**Comment by fulghum** [comment]
Quick update... I've got something rough working for this. The code for this is a little different from `auto_increment`, since `auto_increment` is a property on the schema and this is based on looking at the tuple values/expressions for each row of data being inserted. 

There's a design question around whether nested expressions using `UUID()` should be captured – I'm currently leaning towards only capturing the `UUID()` result and exposing it through `last_insert_uuid()` if the value inserted in the cell is directly from the `UUID()` function without being further transformed. In other words, if it's a complex expression like `CONCAT('foo-', UUID())` we wouldn't capture that value for `last_insert_uuid()`. My reasoning is that the captured UUID wouldn't be directly usable and capturing the result of the complex expression felt a little surprising and wouldn't even necessarily return the same type as `UUID()`. Let me know if that lines up with how you'd like to use the `last_insert_uuid()` function if you have any thoughts there. 

I'll keep working on getting this cleaned up, finishing the tests, and getting a PR out.

**Comment by hannasty** [comment]
Hey @fulghum , thank you for your fast response and work on this topic. I'm working with @gitasmus and we are very impressed by how fast you guys react to our issues. I want to add that using complex expressions is really common when using `UUID()`, because it only returns a `CHAR(36)`. But if you want to store the UUID as a `BINARY(16)` you need to call `UUID_TO_BIN()` also. 
Would it be possible to just store the last generated UUID (e.g. in a session variable) whenever UUID() is called? This of course wouldn't necessarily only apply to inserts, so maybe you could name the procedure something like `last_generated_uuid()` instead of `last_insert_uuid()`. 
Another idea I had, was to supply a second procedure to generate UUIDs directly as a `BINARY(16)`, maybe called 'UUID_BIN()'. This way you could choose if you want a CHAR(36) or a BINARY(16) without the need of complex expressions.

**Comment by fulghum** [comment]
Thanks for the feedback @hannasty. It's very helpful to hear more about how you're using the `UUID()` function and how you want to access the last generated value. 

I like your idea of keeping this really simple and just having the `UUID()` function set a session variable with the last generated value. This makes the implementation a lot simpler, too (e.g. not restricting to only usage in `INSERT`s and following the same behavior as `last_insert_id()`). 

Instead of a new function, I'm picturing just directly accessing the session variable, something like this:
```sql
insert into t values (UUID(), "data");
select @@last_generated_uuid;
```

I'll get started building that out and seeing how it looks.

**Comment by fulghum** [comment]
I implemented this new approach today and spent some time chatting with a teammate about it. We like this feature idea, and we're debating the best interface to expose this information. One advantage of the session variable approach is that it has a very simple implementation. The biggest disadvantage we see is that it's not the same contract as `last_insert_id()`, specifically... it's not guaranteed to be a UUID that was generated for use in a key when a row was inserted. We think there is some benefit for matching the `last_insert_id()` contract for UUID and want to explore that approach a little bit more. One of the concerns was that if a UUID() function was invoked multiple times when inserting a row (e.g. for multiple columns), you wouldn't have a guarantee that we return the right identifier. Having some sort of identifier on the column (analogous to `auto_increment`) is one potential way to ensure we would return the right value.

I'll keep discussing the different approaches with the team. If you have thoughts/opinions, please share, it'd be very helpful for us to make sure we build the right thing for you. For example, it'd be helpful to hear if the callout on multiple `UUID()` function invocations is a potential concern for you or not.

**Comment by hannasty** [comment]
@fulghum 
> If you have thoughts/opinions, please share, it'd be very helpful for us to make sure we build the right thing for you. For example, it'd be helpful to hear if the callout on multiple `UUID()` function invocations is a potential concern for you or not.

In our case it would be no problem because we are generating UUIDs only for primary keys. But as you mentioned you could use `UUID()` more often when inserting a row. For example when working with the JSON type I can imagine that someone wants to generate UUIDs to identify JSON objects. In this case it wouldn't be clear which ID the session variable would contain. So you're right. This can lead to confusion.

@zachmu
> Does that match your expectations for how such a feature should work?

Yes, absolutely. I think this solution would perfectly solve the issue mentioned above. Also it's completely fine that the return type would always be a string as long as the actual column could be a `BINARY(16)`.

Thank you guys for your work on this!

**Comment by fulghum** [comment]
Thanks @hannasty. I've got https://github.com/dolthub/go-mysql-server/pull/2362 opened up for review now, with the `last_insert_uuid()` approach we just discussed. I think this interface turned out pretty nice! Thank you for helping provide feedback while we iterated on it. 🙏  

We'll get this reviewed and released and keep you updated.

**Comment by fulghum** [comment]
Just a quick update that we're getting really close on this one. We found an interesting edge case for `auto_increment` that our current structure couldn't support (https://github.com/dolthub/dolt/issues/7565), so we took a slightly different approach for the auto UUID feature so that it won't have the same limitation. We'll get that reviewed tomorrow and will keep you updated on the progress.

**Comment by fulghum** [comment]
We just released [Dolt 1.35.1](https://github.com/dolthub/dolt/releases/tag/v1.35.1), which contains the new `last_insert_uuid()` function. Thank you @gitasmus and @hannasty for the awesome feature request! 🙏 This was a fun one to build, and I think it's going to be really useful for working with UUIDs. 

I'll update our docs to include this function and we'll probably write a quick blog for it in the next month or so, too. Until then, you can also find the [example usage and requirements for auto UUID columns in the 1.35.1 release notes](https://github.com/dolthub/dolt/releases/tag/v1.35.1). 

Let us know if you have any issues using this new function or if there's anything else we can do to help you build with Dolt!
